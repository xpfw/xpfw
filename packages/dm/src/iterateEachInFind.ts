import { cloneDeep, get, isArray, isObject } from "lodash"

export interface IEachInFindOptions {
  pageSize?: number
}
const iterateEachInFind =
async (collection: string | string[],
       queryObj: any, findMethod: (c: string, q: any) => Promise<any>, iterator: (obj: any) => Promise<any>,
       options?: IEachInFindOptions) => {
  console.log(`entering with collection", collection`)
  if (!isObject(queryObj)) {
    queryObj = {}
  }
  if (collection != null && !Array.isArray(collection)) {
    collection = [collection]
  }
  const pageSize: any = get(options, "pageSize", 5)
  console.log("Trying to iterate", collection)
  for (const col of collection) {
    let gotMorePages = true
    let currentPage = 0
    let res
    while (gotMorePages) {
      const queryWith = cloneDeep(queryObj)
      queryWith.$limit = pageSize
      queryWith.$skip = pageSize * currentPage
      const findRes = await findMethod(col, queryWith)
      if (findRes != null) {
        if (isArray(findRes.data)) {
          for (const e of findRes.data) {
            e.fromCollection = col
            res = await iterator(e)
          }
        }
        currentPage++
        if (currentPage >= findRes.total / pageSize) {
          gotMorePages = false
        }
      } else {
        gotMorePages = false
      }
    }
    return res
  }
}

export default iterateEachInFind
