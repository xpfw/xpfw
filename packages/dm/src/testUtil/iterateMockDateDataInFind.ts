import * as moment from "moment"

const iterateMockDataInFind = (c: string, q: any) => {
  const data1: any = []
  const data2: any = []

  for (let i = 0; i < 60; i++) {
    if (i > 30) {
      data1.push({
        myNum: i * 4 + 2,
        createdAt: moment.utc().add(i - 30, "d").toDate(),
        compDate: moment.utc().subtract(i, "d").toDate()
      })
    } else {
      data2.push({
        myNum: i * 4 + 2,
        createdAt: moment.utc().subtract(i, "d").toDate(),
        compDate: moment.utc().add(i - 30, "d").toDate()
      })
    }
  }
  if (q.$skip === 0) {
    return Promise.resolve({total: 10, data: data1})
  }
  return Promise.resolve({total: 10, data: data2})
}

export default iterateMockDataInFind
