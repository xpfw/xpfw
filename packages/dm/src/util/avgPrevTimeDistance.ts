import { get, isDate } from "lodash"
import { CompareNumberCalculator } from "../typeDef"
import getDate from "./getDate"

/**
 * Calculates the sum of itemPath value for items in
 * If nothing is set returns the sum of iterated items
 * @param config Set itemPath of the number you want the sum of
 */
const avgPrevTimeDistance: CompareNumberCalculator = (config) => {
  const pathToGet: any = get(config, "itemPath", false)
  let total: any = 0
  let amt = 0
  let prev: any = null
  return (obj) => {
    const prevObjDate = getDate(prev, pathToGet)
    const prevIsValid = isDate(prevObjDate)
    if (pathToGet) {
      if (prevIsValid) {
        const objDate = getDate(obj, pathToGet)
        if (isDate(objDate)) {
          total += Math.abs(prevObjDate.getTime() - objDate.getTime())
          amt++
        }
      }
    }
    prev = obj
    if (total === 0) {
      return Promise.resolve(0)
    }
    return Promise.resolve(total / amt)
  }
}

export default avgPrevTimeDistance
