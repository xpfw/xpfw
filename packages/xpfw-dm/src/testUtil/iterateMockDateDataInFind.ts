import * as moment from "moment"
const data1: any = []
const data2: any = []

for (let i = 0; i < 60; i++) {
  if (i > 30) {
    data1.push({
      myNum: i * 4 + 2,
      createdAt: moment().add(i - 30, "d").toDate(),
      compDate: moment().subtract(i, "d").toDate()
    })
  } else {
    data2.push({
      myNum: i * 4 + 2,
      createdAt: moment().subtract(i, "d").toDate(),
      compDate: moment().add(i - 30, "d").toDate()
    })
  }
}

const iterateMockDataInFind = (c: string, q: any) => {
  if (q.$skip === 0) {
    return Promise.resolve({total: 10, data: data1})
  }
  return Promise.resolve({total: 10, data: data2})
}

export default iterateMockDataInFind
