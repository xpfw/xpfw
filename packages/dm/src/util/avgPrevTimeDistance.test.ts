
import MockDate from "mockdate"
MockDate.set(new Date(259542000000))

import iterateEachInFind from "../iterateEachInFind"
import iterateMockDateDataInFind from "../testUtil/iterateMockDateDataInFind"
import avgPrevTimeDistance from "./avgPrevTimeDistance"

test("verify avgprevtimedistance caluclation", async () => {
  expect(await iterateEachInFind("unknown", {}, iterateMockDateDataInFind, avgPrevTimeDistance({
    itemPath: "createdAt"
  })))
    .toMatchSnapshot("configless mean of mockData")
})
