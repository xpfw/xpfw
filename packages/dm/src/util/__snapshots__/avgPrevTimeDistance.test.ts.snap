// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`verify avgprevtimedistance caluclation: configless mean of mockData 1`] = `127403389.83050847`;
